src/MultiOwnerLightAccount.sol

```diff
@@ -0,0 +1,336 @@
+    }
+
+    function _initialize(address[] calldata owners_) internal virtual {
+        emit LightAccountInitialized(_ENTRY_POINT, owners_);
+        _updateOwners(owners_, new address[](0));
+    }
+
+    function _updateOwners(address[] memory ownersToAdd, address[] memory ownersToRemove) internal {
+        _removeOwnersOrRevert(ownersToRemove);
+        _addOwnersOrRevert(ownersToAdd);
+
+        if (_getStorage().owners.isEmpty()) {
+            revert EmptyOwnersNotAllowed();
+        }
+
+        emit OwnersUpdated(ownersToAdd, ownersToRemove);
+    }
+
+    function _addOwnersOrRevert(address[] memory ownersToAdd) internal {
+        LightAccountStorage storage _storage = _getStorage();
+        uint256 length = ownersToAdd.length;
+        for (uint256 i = 0; i < length; ++i) {
+            address ownerToAdd = ownersToAdd[i];
+            if (
+                ownerToAdd == address(0) || ownerToAdd == address(this)
+                    || !_storage.owners.tryAdd(ownerToAdd.toSetValue())
+            ) {
+                revert InvalidOwner(ownerToAdd);
+            }
+        }
+    }
+
+    function _removeOwnersOrRevert(address[] memory ownersToRemove) internal {
+        LightAccountStorage storage _storage = _getStorage();
+        uint256 length = ownersToRemove.length;
+        for (uint256 i = 0; i < length; ++i) {
+            if (!_storage.owners.tryRemove(ownersToRemove[i].toSetValue())) {
+                revert OwnerDoesNotExist(ownersToRemove[i]);
+            }
+        }
+    }
+
+    /// @dev Implement template method of BaseAccount.
+    /// Uses a modified version of `SignatureChecker.isValidSignatureNow` in which the digest is wrapped with an
+    /// "Ethereum Signed Message" envelope for the EOA-owner case but not in the ERC-1271 contract-owner case.
+    function _validateSignature(PackedUserOperation calldata userOp, bytes32 userOpHash)
+        internal
+        virtual
+        override
+        returns (uint256 validationData)
+    {
+        bytes32 signedHash = userOpHash.toEthSignedMessageHash();
+        bytes memory signature = userOp.signature;
+        (address recovered, ECDSA.RecoverError error,) = signedHash.tryRecover(signature);
+        if (error == ECDSA.RecoverError.NoError && _getStorage().owners.contains(recovered.toSetValue())) {
+            return 0;
+        }
+        if (_isValidERC1271SignatureNow(userOpHash, signature)) {
+            return 0;
+        }
+        return SIG_VALIDATION_FAILED;
+    }
+
+    function _isValidERC1271SignatureNow(bytes32 digest, bytes memory signature) internal view returns (bool) {
+        LightAccountStorage storage _storage = _getStorage();
+        address[] memory owners_ = _storage.owners.getAll().toAddressArray();
+        uint256 length = owners_.length;
+        for (uint256 i = 0; i < length; ++i) {
+            if (SignatureChecker.isValidERC1271SignatureNow(owners_[i], digest, signature)) {
+                return true;
+            }
```

Comment:
In a later PR, we should split the signature checking based on a provided address, rather than iterating.